torch/csrc/autograd/engine.h

```diff
@@ -366,7 +366,7 @@ struct TORCH_API Engine {
 
 private:
   // Number of non-reentrant threads
-  std::atomic<uint32_t> non_reentrant_device_thread_count_;
+  volatile uint32_t non_reentrant_device_thread_count_;
```

Comment:
The volatile here is to ensure no sync issues across the mutex?

Reply:
The change looks good, but we should remove the “volatile” qualifier. If all the accesses are protected by the same mutex, then there isn’t a data race. And if there were a data race, the volatile wouldn’t save us (at least not in the abstract model).

Reply:
@colesbury aren't `volatile` ensures that compiler would not optimize:
```
while (!foo) do_bar_that_doesnot_use_foo();
```
to
```
if (!foo)
while(true) do_bar_that_doesnot_use_foo();
```
I.e. https://stackoverflow.com/questions/12878344/volatile-in-c11

Reply:
This code has a mutex, which makes it different from that example (which had no inter-thread synchronization). Our reading code is:

```c++
std::unique_lock<std::mutex> lk(non_reentrant_device_thread_mutex_)
while(non_reentrant_device_thread_count_ != 0) {
    /* NOTE: this unlocks and locks the mutex! */
    non_reentrant_device_thread_condvar_.wait(lk); 
}
```

The relevant writing code for shutdown is:

```c++
std::unique_lock<std::mutex> lk(non_reentrant_device_thread_mutex_);
non_reentrant_device_thread_count_--;
non_reentrant_device_thread_condvar_.notify_one();
``` 

Note that lifting of `non_reentrant_device_thread_count_` out of the while-loop in the reader would also imply reusing the value **across the unlock and lock** operations of `non_reentrant_device_thread_condvar_.wait(lk);`. That's not allowed. The cppreference site has a good explanation [here](https://en.cppreference.com/w/cpp/atomic/memory_order) (emphasis mine):

> Mutual exclusion locks, such as std::mutex or atomic spinlock, are an example of release-acquire synchronization: when the lock is released by thread A and acquired by thread B, **everything that took place in the critical section** (before the release) in the context of thread A **has to be visible to thread B** (after the acquire) which is executing the same critical section.

I think the relevant requirements of the standard ([30.4.1]) are roughly:
a) std::mutex.unlock is a release operation
b) std::mutex.lock is an acquire operation
c) std::mutex.unlock *synchronizes with* subsequent lock operations

There's a sense in which your concern is valid: there's no timeliness guarantee that the reader ever observes the unlock() operation. But that's not really a practical concern and isn't affected by the `volatile` qualifier on `non_reentrant_device_thread_count_`.